<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Data Science Immersive - Capstone Project

---

# Predicting delayed flights and improving operational performance  

## 01. Data Collection and Cleaning  
<br/><br/>

***

## Table of Contents

* [Objectives & Methodology](#Objectives-&-Methodology)
* [1. Data Collection](#1.-Data-Collection)
    * [1.1 Import libraries](#1.1-Import-libraries)
    * [1.2 Import operational data](#1.2-Import-operational-data)
    * [1.3 Import meteorological data](#1.3-Import-meteorological-data)
    * [1.4 Combined Dataset](#1.4-Combined-Dataset)
* [2. Data Cleaning and Processing](#2.-Data-Cleaning-and-Processing)
    * [2.1 Passengers columns](#2.1-Passengers-columns)
    * [2.2 Load Factor column](#2.2-Load-Factor-column)  
<br/><br/>


***

# Objectives & Methodology


This project aims to identify the most important factors leading to a delayed flight, and subsequently provide insights to the airport operator in order to reduce overall delays and improve business performance.  
To achieve this, I analyzed the operational database of a major international airport containing all the flights made in 2013 after merging it with hourly meteorogical data from the Open-Meteo website.  
In the final stage, I created classification models to determine which factors can have the greatest impact on flight delays.  
<br/><br/>

# 1. Data Collection

### 1.1 Import libraries

In [1]:
import pandas as pd
import numpy as np
import warnings

In [2]:
# run to avoid warnings
warnings.filterwarnings('ignore')

### 1.2 Import operational data

In [3]:
# create a dataframe from the original operational database
odb = pd.read_excel('datasets\_original\odb\AODB2013.xlsx', 'AODB2013')
odb

Actual Year  Actual Month  Actual Day  Actual Hour  D/A Terminal Num  \
0              2013             1           1            3  ARR            2   
1              2013             1           1            3  ARR            1   
2              2013             1           1            5  DEP            2   
3              2013             1           1            6  DEP            1   
4              2013             1           1            7  ARR            2   
...             ...           ...         ...          ...  ...          ...   
195288         2013            12          31           22  ARR            1   
195289         2013            12          31           22  ARR            1   
195290         2013            12          31           22  ARR            1   
195291         2013            12          31           22  DEP            2   
195292         2013            12          31           23  ARR            3   

       Airline  Code                  Airline Name AirCraft Type Code  \
0                 N4          ООО "СЕВЕРНЫЙ ВЕТЕР"                763   
1                 OZ               ASIANA AIRLINES                74Y   
2                 N4          ООО "СЕВЕРНЫЙ ВЕТЕР"                763   
3                 OZ               ASIANA AIRLINES                74Y   
4                 LY     EL AL ISRAEL AIRLINES LTD                738   
...              ...                           ...                ...   
195288            FV    ОАО "АВИАКОМПАНИЯ "РОССИЯ"                A81   
195289            FV    ОАО "АВИАКОМПАНИЯ "РОССИЯ"                A81   
195290            U6  ОАО АК "УРАЛЬСКИЕ АВИАЛИНИИ"                321   
195291            BT                    AIR BALTIC                735   
195292           NaN                           NaN                D2T   

       AirCraft Type Name   ... Pax 1 Pax 2 Pax 3 Pax 4 Pax 5 Pax 6  \
0                   BOEING  ...   196     0     0     0     0     0   
1                   BOEING  ...     0     0     0     0     0     0   
2                   BOEING  ...   258    26     2     0     4   280   
3                   BOEING  ...     0     0     0     0     0     0   
4                   BOEING  ...    13     1     1     0     0    14   
...                    ...  ...   ...   ...   ...   ...   ...   ...   
195288             ANTONOV  ...    26     0     0     0     0     0   
195289             ANTONOV  ...    44     2     1     0     0     0   
195290    AIRBUS INDUSTRIE  ...   174    24     3     0     0   198   
195291              BOEING  ...    73     0     0     0     0    73   
195292            DASSAULT  ...     0     0     0     0     0     0   

       Flight Count Load Percent Baggage Weight Mail Weight  
0                 1       67.586         3327.0         NaN  
1                 1          NaN            NaN         NaN  
2                 1       94.667         3224.0         0.0  
3                 1          NaN            NaN         NaN  
4                 1        8.235          244.0       173.0  
...             ...          ...            ...         ...  
195288            1       34.667          211.0         0.0  
195289            1       61.333          355.0         0.0  
195290            1      104.211         2044.0         0.0  
195291            1       64.035          334.0       637.0  
195292            1      100.000            NaN         NaN  

[195293 rows x 32 columns]

In [4]:
odb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195293 entries, 0 to 195292
Data columns (total 32 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Actual Year          195293 non-null  int64  
 1   Actual Month         195293 non-null  int64  
 2   Actual Day           195293 non-null  int64  
 3   Actual Hour          195293 non-null  int64  
 4   D/A                  195293 non-null  object 
 5   Terminal Num         135431 non-null  object 
 6   Airline  Code        184132 non-null  object 
 7   Airline Name         183940 non-null  object 
 8   AirCraft Type Code   137357 non-null  object 
 9   AirCraft Type Name   137357 non-null  object 
 10  Service Type         195293 non-null  object 
 11  Direction Name       195292 non-null  object 
 12  Cancellation         89 non-null      object 
 13  Regularity           195293 non-null  object 
 14  CIS Ind              20733 non-null   object 
 15  Country Name Dire

In [5]:
# check for missing values
odb.isnull().sum()

Actual Year                 0
Actual Month                0
Actual Day                  0
Actual Hour                 0
D/A                         0
Terminal Num            59862
Airline  Code           11161
Airline Name            11353
AirCraft Type Code      57936
AirCraft Type Name      57936
Service Type                0
Direction Name              1
Cancellation           195204
Regularity                  0
CIS Ind                174560
Country Name Direct         0
Flight Number               4
Main Delay Code        162033
Source Delay           180400
RunWay                 126803
Spot Number             57927
Pax Total               72275
Pax 1                       0
Pax 2                       0
Pax 3                       0
Pax 4                       0
Pax 5                       0
Pax 6                       0
Flight Count                0
Load Percent            12846
Baggage Weight          74207
Mail Weight            112335
dtype: int64

There are several columns with missing values, they will be addressed in different ways before the modelling stage.


#### Assumption

My first assumption is that a non-null value in the **'Main Delay Code'** column signals a delay.  
We have 33260 non-null values over 195293 total rows, so this means that delayed flights account for roughly 17% of the total flights.  
This figure seems in line with industry average.  
<br/><br/>

In [6]:
# check the values percentages in the 'Main Delay Code' column
odb['Main Delay Code'].value_counts(normalize=True)

93     0.317108
99     0.137462
89     0.075135
3      0.057005
85     0.049880
         ...   
27     0.000060
25     0.000060
60     0.000060
93W    0.000030
56     0.000030
Name: Main Delay Code, Length: 88, dtype: float64

In [7]:
# confirm the sum of the different values
odb['Main Delay Code'].value_counts().sum()

33260

In [8]:
# display the unique values in the column
odb['Main Delay Code'].unique()

array([nan, 89, 86, 15, 75, 93, 4, 46, 3, 14, 85, 99, 36, 10, 72, 18, 63,
       41, 11, 32, 5, 47, 96, 81, 87, 34, 77, 65, 31, 1, 95, 91, 43, 33,
       38, 39, 16, 29, 97, 66, 42, 19, 37, 6, 94, 68, 88, 12, 2, 92, 9,
       61, 76, 82, 51, 23, 69, 55, 71, 35, 8, 57, 62, 22, 67, 21, 70, 25,
       0, 98, 13, 27, 45, 58, 52, 28, 64, 83, 49, 17, 84, 44, 73, 60,
       '93W', 48, 24, 56, '93Y'], dtype=object)

In [9]:
# check the values percentages in the 'Source Delay' column
odb['Source Delay'].value_counts(normalize=True)

А/К     0.714295
ВНЕШ    0.265561
А/П     0.020144
Name: Source Delay, dtype: float64

In [10]:
# confirm the sum of the different values
odb['Source Delay'].value_counts().sum()

14893

#### Assumption

The **'Main Delay Code'** column has significantly more values than the **'Source Delay'** column, I assume the latter is not complete.  
I don't have any additional information about the different delay code values, so I treat them all the same.  
<br/><br/>

In [11]:
# display the statistical summary of the numerical features
odb.describe()

Actual Year   Actual Month     Actual Day    Actual Hour  \
count     195293.0  195293.000000  195293.000000  195293.000000   
mean        2013.0       7.246266      16.024609      13.410388   
std            0.0       3.082975       8.765956       6.164239   
min         2013.0       1.000000       1.000000       0.000000   
25%         2013.0       5.000000       9.000000       9.000000   
50%         2013.0       7.000000      16.000000      14.000000   
75%         2013.0      10.000000      24.000000      18.000000   
max         2013.0      12.000000      31.000000      23.000000   

           Pax Total          Pax 1          Pax 2          Pax 3  \
count  123018.000000  195293.000000  195293.000000  195293.000000   
mean      103.844649      62.046279       3.360161       0.720062   
std        65.204536      66.526937       8.729913       2.025564   
min         0.000000       0.000000       0.000000       0.000000   
25%        63.000000       0.000000       0.000000       0.000000   
50%        98.000000      54.000000       0.000000       0.000000   
75%       130.000000     107.000000       3.000000       1.000000   
max      1002.000000     999.000000     425.000000     337.000000   

               Pax 4          Pax 5          Pax 6  Flight Count  \
count  195293.000000  195293.000000  195293.000000      195293.0   
mean        0.046453       2.024307      56.529748           1.0   
std         1.152120       5.212153      69.956528           0.0   
min         0.000000       0.000000       0.000000           1.0   
25%         0.000000       0.000000       0.000000           1.0   
50%         0.000000       0.000000      38.000000           1.0   
75%         0.000000       2.000000     103.000000           1.0   
max       298.000000     479.000000     605.000000           1.0   

        Load Percent  Baggage Weight   Mail Weight  
count  182447.000000   121086.000000  82958.000000  
mean       81.163451     1134.466371     50.708612  
std        41.278123     1015.762045    140.169544  
min         0.000000        0.000000      0.000000  
25%        66.026000      458.000000      0.000000  
50%        89.474000      887.000000      0.000000  
75%       100.000000     1512.000000     23.000000  
max     13800.000000    29117.000000   3842.000000

In [12]:
# update column names
odb.columns = ['year', 'month', 'day', 'hour', 'dep_arr',
       'terminal_num', 'airline_code', 'airline_name', 'aircraft_type_code',
       'aircraft_type_name', 'service_type', 'dom_int', 'cancellation',
       'regularity', 'CIS_ind', 'country_name', 'flight_number',
       'main_delay_code', 'source_delay', 'runway', 'spot_number', 'pax_total',
       'pax_1', 'pax_2', 'pax_3', 'pax_4', 'pax_5', 'pax_6', 'flight_count',
       'load_factor', 'baggage_weight', 'mail_weight']

In [13]:
odb

year  month  day  hour dep_arr terminal_num airline_code  \
0       2013      1    1     3     ARR            2           N4   
1       2013      1    1     3     ARR            1           OZ   
2       2013      1    1     5     DEP            2           N4   
3       2013      1    1     6     DEP            1           OZ   
4       2013      1    1     7     ARR            2           LY   
...      ...    ...  ...   ...     ...          ...          ...   
195288  2013     12   31    22     ARR            1           FV   
195289  2013     12   31    22     ARR            1           FV   
195290  2013     12   31    22     ARR            1           U6   
195291  2013     12   31    22     DEP            2           BT   
195292  2013     12   31    23     ARR            3          NaN   

                        airline_name aircraft_type_code aircraft_type_name  \
0               ООО "СЕВЕРНЫЙ ВЕТЕР"                763             BOEING   
1                    ASIANA AIRLINES                74Y             BOEING   
2               ООО "СЕВЕРНЫЙ ВЕТЕР"                763             BOEING   
3                    ASIANA AIRLINES                74Y             BOEING   
4          EL AL ISRAEL AIRLINES LTD                738             BOEING   
...                              ...                ...                ...   
195288    ОАО "АВИАКОМПАНИЯ "РОССИЯ"                A81            ANTONOV   
195289    ОАО "АВИАКОМПАНИЯ "РОССИЯ"                A81            ANTONOV   
195290  ОАО АК "УРАЛЬСКИЕ АВИАЛИНИИ"                321   AIRBUS INDUSTRIE   
195291                    AIR BALTIC                735             BOEING   
195292                           NaN                D2T           DASSAULT   

        ... pax_1 pax_2 pax_3 pax_4 pax_5 pax_6 flight_count load_factor  \
0       ...   196     0     0     0     0     0            1      67.586   
1       ...     0     0     0     0     0     0            1         NaN   
2       ...   258    26     2     0     4   280            1      94.667   
3       ...     0     0     0     0     0     0            1         NaN   
4       ...    13     1     1     0     0    14            1       8.235   
...     ...   ...   ...   ...   ...   ...   ...          ...         ...   
195288  ...    26     0     0     0     0     0            1      34.667   
195289  ...    44     2     1     0     0     0            1      61.333   
195290  ...   174    24     3     0     0   198            1     104.211   
195291  ...    73     0     0     0     0    73            1      64.035   
195292  ...     0     0     0     0     0     0            1     100.000   

       baggage_weight mail_weight  
0              3327.0         NaN  
1                 NaN         NaN  
2              3224.0         0.0  
3                 NaN         NaN  
4               244.0       173.0  
...               ...         ...  
195288          211.0         0.0  
195289          355.0         0.0  
195290         2044.0         0.0  
195291          334.0       637.0  
195292            NaN         NaN  

[195293 rows x 32 columns]

In [14]:
# create datetime column
odb['date'] = pd.to_datetime(odb[['year', 'month', 'day', 'hour']])

# create day-of-week column
odb['weekday'] = odb['date'].dt.dayofweek

# remove seconds from datetime column
odb['date'] = odb['date'].apply(lambda t: t.strftime('%Y-%m-%d %H:%M'))

# rearrange columns order
odb = odb[['date', 'year', 'month', 'day', 'weekday', 'hour', 'dep_arr',
       'terminal_num', 'airline_code', 'airline_name', 'aircraft_type_code',
       'aircraft_type_name', 'service_type', 'dom_int', 'cancellation',
       'regularity', 'CIS_ind', 'country_name', 'flight_number',
       'main_delay_code', 'source_delay', 'runway', 'spot_number', 'pax_total',
       'pax_1', 'pax_2', 'pax_3', 'pax_4', 'pax_5', 'pax_6', 'flight_count',
       'load_factor', 'baggage_weight', 'mail_weight']]

In [15]:
odb

date  year  month  day  weekday  hour dep_arr  \
0       2013-01-01 03:00  2013      1    1        1     3     ARR   
1       2013-01-01 03:00  2013      1    1        1     3     ARR   
2       2013-01-01 05:00  2013      1    1        1     5     DEP   
3       2013-01-01 06:00  2013      1    1        1     6     DEP   
4       2013-01-01 07:00  2013      1    1        1     7     ARR   
...                  ...   ...    ...  ...      ...   ...     ...   
195288  2013-12-31 22:00  2013     12   31        1    22     ARR   
195289  2013-12-31 22:00  2013     12   31        1    22     ARR   
195290  2013-12-31 22:00  2013     12   31        1    22     ARR   
195291  2013-12-31 22:00  2013     12   31        1    22     DEP   
195292  2013-12-31 23:00  2013     12   31        1    23     ARR   

       terminal_num airline_code                  airline_name  ... pax_1  \
0                 2           N4          ООО "СЕВЕРНЫЙ ВЕТЕР"  ...   196   
1                 1           OZ               ASIANA AIRLINES  ...     0   
2                 2           N4          ООО "СЕВЕРНЫЙ ВЕТЕР"  ...   258   
3                 1           OZ               ASIANA AIRLINES  ...     0   
4                 2           LY     EL AL ISRAEL AIRLINES LTD  ...    13   
...             ...          ...                           ...  ...   ...   
195288            1           FV    ОАО "АВИАКОМПАНИЯ "РОССИЯ"  ...    26   
195289            1           FV    ОАО "АВИАКОМПАНИЯ "РОССИЯ"  ...    44   
195290            1           U6  ОАО АК "УРАЛЬСКИЕ АВИАЛИНИИ"  ...   174   
195291            2           BT                    AIR BALTIC  ...    73   
195292            3          NaN                           NaN  ...     0   

       pax_2 pax_3 pax_4 pax_5 pax_6 flight_count load_factor baggage_weight  \
0          0     0     0     0     0            1      67.586         3327.0   
1          0     0     0     0     0            1         NaN            NaN   
2         26     2     0     4   280            1      94.667         3224.0   
3          0     0     0     0     0            1         NaN            NaN   
4          1     1     0     0    14            1       8.235          244.0   
...      ...   ...   ...   ...   ...          ...         ...            ...   
195288     0     0     0     0     0            1      34.667          211.0   
195289     2     1     0     0     0            1      61.333          355.0   
195290    24     3     0     0   198            1     104.211         2044.0   
195291     0     0     0     0    73            1      64.035          334.0   
195292     0     0     0     0     0            1     100.000            NaN   

       mail_weight  
0              NaN  
1              NaN  
2              0.0  
3              NaN  
4            173.0  
...            ...  
195288         0.0  
195289         0.0  
195290         0.0  
195291       637.0  
195292         NaN  

[195293 rows x 34 columns]

### 1.3 Import meteorological data

In [16]:
# create a dataframe from the original weather dataset
weather = pd.read_csv(r'datasets\base\open-meteo-59-80N30-30E19m_mod.csv')
weather

time  temperature_2m (°C)  relativehumidity_2m (%)  \
0     2013-01-01T00:00                  0.9                       93   
1     2013-01-01T01:00                  0.9                       93   
2     2013-01-01T02:00                  1.0                       93   
3     2013-01-01T03:00                  0.9                       94   
4     2013-01-01T04:00                  0.9                       94   
...                ...                  ...                      ...   
8755  2013-12-31T19:00                  1.7                       85   
8756  2013-12-31T20:00                  1.4                       85   
8757  2013-12-31T21:00                  1.1                       85   
8758  2013-12-31T22:00                  1.2                       84   
8759  2013-12-31T23:00                  1.2                       83   

      apparent_temperature (°C)  precipitation (mm)  weathercode (wmo code)  \
0                          -2.5                 0.0                       3   
1                          -2.3                 0.0                       3   
2                          -2.3                 0.0                       3   
3                          -2.4                 0.0                       3   
4                          -2.6                 0.0                       3   
...                         ...                 ...                     ...   
8755                       -2.2                 0.0                       2   
8756                       -2.8                 0.0                       2   
8757                       -2.9                 0.0                       2   
8758                       -2.8                 0.0                       2   
8759                       -2.9                 0.0                       2   

      windspeed_10m (km/h)  
0                     12.3  
1                     12.6  
2                     12.6  
3                     13.0  
4                     15.2  
...                    ...  
8755                  16.3  
8756                  16.3  
8757                  15.9  
8758                  16.1  
8759                  16.1  

[8760 rows x 7 columns]

In [17]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   time                       8760 non-null   object 
 1   temperature_2m (°C)        8760 non-null   float64
 2   relativehumidity_2m (%)    8760 non-null   int64  
 3   apparent_temperature (°C)  8760 non-null   float64
 4   precipitation (mm)         8760 non-null   float64
 5   weathercode (wmo code)     8760 non-null   int64  
 6   windspeed_10m (km/h)       8760 non-null   float64
dtypes: float64(4), int64(2), object(1)
memory usage: 479.2+ KB


In [18]:
# check for missing values
weather.isnull().sum()

time                         0
temperature_2m (°C)          0
relativehumidity_2m (%)      0
apparent_temperature (°C)    0
precipitation (mm)           0
weathercode (wmo code)       0
windspeed_10m (km/h)         0
dtype: int64

In [19]:
# update column names
weather.columns = ['date', 'temperature', 'relative_humidity',
       'apparent_temperature', 'precipitation',
       'weather_code', 'wind_speed']

In [20]:
# reformat the 'date' column to match the one in the odb dataframe
weather['date'] = weather['date'].str.replace('T', ' ')

In [21]:
weather

date  temperature  relative_humidity  apparent_temperature  \
0     2013-01-01 00:00          0.9                 93                  -2.5   
1     2013-01-01 01:00          0.9                 93                  -2.3   
2     2013-01-01 02:00          1.0                 93                  -2.3   
3     2013-01-01 03:00          0.9                 94                  -2.4   
4     2013-01-01 04:00          0.9                 94                  -2.6   
...                ...          ...                ...                   ...   
8755  2013-12-31 19:00          1.7                 85                  -2.2   
8756  2013-12-31 20:00          1.4                 85                  -2.8   
8757  2013-12-31 21:00          1.1                 85                  -2.9   
8758  2013-12-31 22:00          1.2                 84                  -2.8   
8759  2013-12-31 23:00          1.2                 83                  -2.9   

      precipitation  weather_code  wind_speed  
0               0.0             3        12.3  
1               0.0             3        12.6  
2               0.0             3        12.6  
3               0.0             3        13.0  
4               0.0             3        15.2  
...             ...           ...         ...  
8755            0.0             2        16.3  
8756            0.0             2        16.3  
8757            0.0             2        15.9  
8758            0.0             2        16.1  
8759            0.0             2        16.1  

[8760 rows x 7 columns]

### 1.4 Combined Dataset

In [22]:
# merge the two dataframes using the 'date' column as key
df = pd.merge(odb, weather, how='left', on='date')

In [23]:
df

date  year  month  day  weekday  hour dep_arr  \
0       2013-01-01 03:00  2013      1    1        1     3     ARR   
1       2013-01-01 03:00  2013      1    1        1     3     ARR   
2       2013-01-01 05:00  2013      1    1        1     5     DEP   
3       2013-01-01 06:00  2013      1    1        1     6     DEP   
4       2013-01-01 07:00  2013      1    1        1     7     ARR   
...                  ...   ...    ...  ...      ...   ...     ...   
195288  2013-12-31 22:00  2013     12   31        1    22     ARR   
195289  2013-12-31 22:00  2013     12   31        1    22     ARR   
195290  2013-12-31 22:00  2013     12   31        1    22     ARR   
195291  2013-12-31 22:00  2013     12   31        1    22     DEP   
195292  2013-12-31 23:00  2013     12   31        1    23     ARR   

       terminal_num airline_code                  airline_name  ...  \
0                 2           N4          ООО "СЕВЕРНЫЙ ВЕТЕР"  ...   
1                 1           OZ               ASIANA AIRLINES  ...   
2                 2           N4          ООО "СЕВЕРНЫЙ ВЕТЕР"  ...   
3                 1           OZ               ASIANA AIRLINES  ...   
4                 2           LY     EL AL ISRAEL AIRLINES LTD  ...   
...             ...          ...                           ...  ...   
195288            1           FV    ОАО "АВИАКОМПАНИЯ "РОССИЯ"  ...   
195289            1           FV    ОАО "АВИАКОМПАНИЯ "РОССИЯ"  ...   
195290            1           U6  ОАО АК "УРАЛЬСКИЕ АВИАЛИНИИ"  ...   
195291            2           BT                    AIR BALTIC  ...   
195292            3          NaN                           NaN  ...   

       flight_count load_factor baggage_weight mail_weight temperature  \
0                 1      67.586         3327.0         NaN         0.9   
1                 1         NaN            NaN         NaN         0.9   
2                 1      94.667         3224.0         0.0         0.9   
3                 1         NaN            NaN         NaN         0.7   
4                 1       8.235          244.0       173.0         0.8   
...             ...         ...            ...         ...         ...   
195288            1      34.667          211.0         0.0         1.2   
195289            1      61.333          355.0         0.0         1.2   
195290            1     104.211         2044.0         0.0         1.2   
195291            1      64.035          334.0       637.0         1.2   
195292            1     100.000            NaN         NaN         1.2   

       relative_humidity apparent_temperature precipitation weather_code  \
0                     94                 -2.4           0.0            3   
1                     94                 -2.4           0.0            3   
2                     94                 -2.9           0.0            3   
3                     94                 -3.3           0.0            3   
4                     94                 -3.6           0.2           71   
...                  ...                  ...           ...          ...   
195288                84                 -2.8           0.0            2   
195289                84                 -2.8           0.0            2   
195290                84                 -2.8           0.0            2   
195291                84                 -2.8           0.0            2   
195292                83                 -2.9           0.0            2   

       wind_speed  
0            13.0  
1            13.0  
2            17.0  
3            18.1  
4            19.6  
...           ...  
195288       16.1  
195289       16.1  
195290       16.1  
195291       16.1  
195292       16.1  

[195293 rows x 40 columns]

In [24]:
# save the modified dataframe as CSV
df.to_csv('datasets/mod/df_00.csv', encoding='utf-8', index=False)

In [25]:
# load data from CSV for consistency reasons
df = pd.read_csv('datasets/mod/df_00.csv', low_memory=False)

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195293 entries, 0 to 195292
Data columns (total 40 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   date                  195293 non-null  object 
 1   year                  195293 non-null  int64  
 2   month                 195293 non-null  int64  
 3   day                   195293 non-null  int64  
 4   weekday               195293 non-null  int64  
 5   hour                  195293 non-null  int64  
 6   dep_arr               195293 non-null  object 
 7   terminal_num          135431 non-null  object 
 8   airline_code          184132 non-null  object 
 9   airline_name          183940 non-null  object 
 10  aircraft_type_code    137357 non-null  object 
 11  aircraft_type_name    137357 non-null  object 
 12  service_type          195293 non-null  object 
 13  dom_int               195292 non-null  object 
 14  cancellation          89 non-null      object 
 15  

In [27]:
# display the size of the dataframe
df.shape

(195293, 40)

# 2. Data Cleaning and Processing

In [28]:
# rename columns with binary values
df.rename(columns={'dep_arr': 'departing', 'dom_int': 'domestic', 'cancellation': 'cancelled',
                   'regularity': 'regular', 'CIS_ind': 'CIS_country'}, inplace=True)

In [29]:
# check the values percentages in the 'departing' column
df.departing.value_counts(normalize=True)

DEP    0.502
ARR    0.498
Name: departing, dtype: float64

In [30]:
# confirm the sum of the different values
df.departing.value_counts().sum()

195293

Departing and arriving flights are almost perfectly balanced, this makes sense from an operational point of view.  
The **'departing'** column has no empty values and will be binarised.  
<br/><br/>

In [31]:
# convert 'departing' column into binary
df.departing[df.departing == 'DEP'] = 1
df.departing[df.departing == 'ARR'] = 0
df

date  year  month  day  weekday  hour departing  \
0       2013-01-01 03:00  2013      1    1        1     3         0   
1       2013-01-01 03:00  2013      1    1        1     3         0   
2       2013-01-01 05:00  2013      1    1        1     5         1   
3       2013-01-01 06:00  2013      1    1        1     6         1   
4       2013-01-01 07:00  2013      1    1        1     7         0   
...                  ...   ...    ...  ...      ...   ...       ...   
195288  2013-12-31 22:00  2013     12   31        1    22         0   
195289  2013-12-31 22:00  2013     12   31        1    22         0   
195290  2013-12-31 22:00  2013     12   31        1    22         0   
195291  2013-12-31 22:00  2013     12   31        1    22         1   
195292  2013-12-31 23:00  2013     12   31        1    23         0   

       terminal_num airline_code                  airline_name  ...  \
0                 2           N4          ООО "СЕВЕРНЫЙ ВЕТЕР"  ...   
1                 1           OZ               ASIANA AIRLINES  ...   
2                 2           N4          ООО "СЕВЕРНЫЙ ВЕТЕР"  ...   
3                 1           OZ               ASIANA AIRLINES  ...   
4                 2           LY     EL AL ISRAEL AIRLINES LTD  ...   
...             ...          ...                           ...  ...   
195288            1           FV    ОАО "АВИАКОМПАНИЯ "РОССИЯ"  ...   
195289            1           FV    ОАО "АВИАКОМПАНИЯ "РОССИЯ"  ...   
195290            1           U6  ОАО АК "УРАЛЬСКИЕ АВИАЛИНИИ"  ...   
195291            2           BT                    AIR BALTIC  ...   
195292            3          NaN                           NaN  ...   

       flight_count load_factor baggage_weight mail_weight temperature  \
0                 1      67.586         3327.0         NaN         0.9   
1                 1         NaN            NaN         NaN         0.9   
2                 1      94.667         3224.0         0.0         0.9   
3                 1         NaN            NaN         NaN         0.7   
4                 1       8.235          244.0       173.0         0.8   
...             ...         ...            ...         ...         ...   
195288            1      34.667          211.0         0.0         1.2   
195289            1      61.333          355.0         0.0         1.2   
195290            1     104.211         2044.0         0.0         1.2   
195291            1      64.035          334.0       637.0         1.2   
195292            1     100.000            NaN         NaN         1.2   

       relative_humidity apparent_temperature precipitation weather_code  \
0                     94                 -2.4           0.0            3   
1                     94                 -2.4           0.0            3   
2                     94                 -2.9           0.0            3   
3                     94                 -3.3           0.0            3   
4                     94                 -3.6           0.2           71   
...                  ...                  ...           ...          ...   
195288                84                 -2.8           0.0            2   
195289                84                 -2.8           0.0            2   
195290                84                 -2.8           0.0            2   
195291                84                 -2.8           0.0            2   
195292                83                 -2.9           0.0            2   

       wind_speed  
0            13.0  
1            13.0  
2            17.0  
3            18.1  
4            19.6  
...           ...  
195288       16.1  
195289       16.1  
195290       16.1  
195291       16.1  
195292       16.1  

[195293 rows x 40 columns]

In [32]:
# check the value percentages in the 'terminal_num' column
df.terminal_num.value_counts(normalize=True)

1                      0.666169
2                      0.292732
3                      0.034490
A                      0.004932
4                      0.001661
2014-02-01 00:00:00    0.000007
2014-01-02 00:00:00    0.000007
Name: terminal_num, dtype: float64

In [33]:
# check the values in the 'terminal_num' column
df.terminal_num.value_counts()

1                      90220
2                      39645
3                       4671
A                        668
4                        225
2014-02-01 00:00:00        1
2014-01-02 00:00:00        1
Name: terminal_num, dtype: int64

In [35]:
# remove incorrect values in the 'terminal_num' column
df.loc[df['terminal_num'] == '2014-02-01 00:00:00', 'terminal_num'] = np.nan
df.loc[df['terminal_num'] == '2014-01-02 00:00:00', 'terminal_num'] = np.nan

In [36]:
df.terminal_num.value_counts()

1    90220
2    39645
3     4671
A      668
4      225
Name: terminal_num, dtype: int64

In [37]:
df.shape

(195293, 40)

In [38]:
# check the value percentages in the 'airline_code' column
df.airline_code.value_counts(normalize=True)

FV     0.320721
SU     0.253291
UN     0.048297
UT     0.025433
S7     0.023581
         ...   
AHO    0.000005
IFA    0.000005
ЕШ     0.000005
HYP    0.000005
VTK    0.000005
Name: airline_code, Length: 207, dtype: float64

In [39]:
# check the value percentages in the 'airline_name' column
df.airline_name.value_counts(normalize=True)

ОАО "АВИАКОМПАНИЯ "РОССИЯ"                 0.321056
ОАО "АЭРОФЛОТ-РОССИЙСКИЕ АВИАЛИНИИ"        0.253556
АВИАКОМПАНИЯ "ТРАНСАЭРО"                   0.048347
ОАО "АВИАКОМПАНИЯ "ЮТЭЙР"                  0.025459
ОАО АВИАКОМПАНИЯ "СИБИРЬ" (S7-AIRLINES)    0.023606
                                             ...   
AIR HAMBURG PRIVATE JETS                   0.000005
ООО "АРМАВИА"                              0.000005
FAI RENT-A-JET                             0.000005
АМУРСКАЯ БАЗА АВИАЛЕСООХРАНЫ               0.000005
ОАО "АК "ВОСТОК"                           0.000005
Name: airline_name, Length: 206, dtype: float64

In [40]:
# display the unique values in the 'airline_name' column
list(df.airline_name.unique())

['ООО "СЕВЕРНЫЙ ВЕТЕР"',
 'ASIANA AIRLINES',
 'EL AL ISRAEL AIRLINES LTD',
 'ОАО "АВИАКОМПАНИЯ "РОССИЯ"',
 'ОАО "ОРЕНБУРГСКИЕ АВИАЛИНИИ"',
 nan,
 'ESTONIAN AIR REGIONAL',
 'AIR BALTIC',
 'FINNAIR',
 'ОАО "АЭРОФЛОТ-РОССИЙСКИЕ АВИАЛИНИИ"',
 'МОСКОВСКИЙ ФИЛИАЛ ОАО АК "ТАЙМЫР" (NORD STAR)',
 'ОАО АК "УРАЛЬСКИЕ АВИАЛИНИИ"',
 'АВИАКОМПАНИЯ "ТРАНСАЭРО"',
 'ООО "ГЛОБУС" (S7 GROUP)',
 'ООО "АВИАКОМПАНИЯ КОГАЛЫМАВИА"',
 'SWISS',
 'EMIRATES AIRLINES',
 'BRITISH AIRWAYS',
 'CZECH AIRLINES',
 'AIR FRANCE',
 'AUSTRIAN AIRLINES',
 'LUFTHANSA GERMAN AIRLINES',
 'AIR BERLIN',
 'SAS-SCANDINAVIAN AIRLINES',
 'KLM-ROYAL DUTCH AIRLINES',
 'ОАО АВИАКОМПАНИЯ "СИБИРЬ" (S7-AIRLINES)',
 'ЗАО "НОРДАВИА-РА"',
 'НАК БЕЛОРУССИИ "БЕЛАВИА"',
 'KOREAN AIR LINES CO.LTD.',
 'AIR ASTANA',
 'ОАО "АВИАКОМПАНИЯ "ЮТЭЙР"',
 'AIR EUROPE (AIR ESPANA)',
 'ОАО "АК БАРС АЭРО"',
 'ЗАО АВИАКОМПАНИЯ "ПОЛЕТ"',
 'BLUE JET',
 'NIKI',
 'UZBEKISTAN AIRWAYS',
 'ОАО "ДОНАВИА"',
 'ЗАО "АЗЕРБАЙДЖАН ХАВА ЙОЛЛАРЫ"',
 'LOT POLISH AIRLINES',
 'AL

Many **airline names** are in the Cyrillic alphabet, they will be translated into English and cleaned up.  
<br/><br/>

In [41]:
# create a dictionary containing the correct airline names
airlines = {
 'ООО "СЕВЕРНЫЙ ВЕТЕР"' : 'NORDWIND AIRLINES', 
 'ASIANA AIRLINES' : 'ASIANA AIRLINES', 
 'EL AL ISRAEL AIRLINES LTD' : 'EL AL ISRAEL AIRLINES', 
 'ОАО "АВИАКОМПАНИЯ "РОССИЯ"' : 'ROSSIYA AIRLINES', 
 'ОАО "ОРЕНБУРГСКИЕ АВИАЛИНИИ"' : 'ORENAIR', 
 'nan' : 'nan', 
 'ESTONIAN AIR REGIONAL' : 'ESTONIAN AIR', 
 'AIR BALTIC' : 'AIRBALTIC', 
 'FINNAIR' : 'FINNAIR', 
 'ОАО "АЭРОФЛОТ-РОССИЙСКИЕ АВИАЛИНИИ"' : 'AEROFLOT', 
 'МОСКОВСКИЙ ФИЛИАЛ ОАО АК "ТАЙМЫР" (NORD STAR)' : 'NORDSTAR', 
 'ОАО АК "УРАЛЬСКИЕ АВИАЛИНИИ"' : 'URAL AIRLINES', 
 'АВИАКОМПАНИЯ "ТРАНСАЭРО"' : 'TRANSAERO', 
 'ООО "ГЛОБУС" (S7 GROUP)' : 'GLOBUS AIRLINES', 
 'ООО "АВИАКОМПАНИЯ КОГАЛЫМАВИА"' : 'METROJET', 
 'SWISS' : 'SWISS', 
 'EMIRATES AIRLINES' : 'EMIRATES', 
 'BRITISH AIRWAYS' : 'BRITISH AIRWAYS', 
 'CZECH AIRLINES' : 'CZECH AIRLINES', 
 'AIR FRANCE' : 'AIR FRANCE', 
 'AUSTRIAN AIRLINES' : 'AUSTRIAN AIRLINES', 
 'LUFTHANSA GERMAN AIRLINES' : 'LUFTHANSA', 
 'AIR BERLIN' : 'AIR BERLIN', 
 'SAS-SCANDINAVIAN AIRLINES' : 'SAS-SCANDINAVIAN AIRLINES', 
 'KLM-ROYAL DUTCH AIRLINES' : 'KLM-ROYAL DUTCH AIRLINES', 
 'ОАО АВИАКОМПАНИЯ "СИБИРЬ" (S7-AIRLINES)' : 'S7 AIRLINES', 
 'ЗАО "НОРДАВИА-РА"' : 'NORDAVIA', 
 'НАК БЕЛОРУССИИ "БЕЛАВИА"' : 'BELAVIA', 
 'KOREAN AIR LINES CO.LTD.' : 'KOREAN AIR', 
 'AIR ASTANA' : 'AIR ASTANA', 
 'ОАО "АВИАКОМПАНИЯ "ЮТЭЙР"' : 'UTAIR', 
 'AIR EUROPE (AIR ESPANA)' : 'AIR EUROPA', 
 'ОАО "АК БАРС АЭРО"' : 'AK BARS AERO', 
 'ЗАО АВИАКОМПАНИЯ "ПОЛЕТ"' : 'POLET AIRLINES', 
 'BLUE JET' : 'BLUE JET', 
 'NIKI' : 'NIKI', 
 'UZBEKISTAN AIRWAYS' : 'UZBEKISTAN AIRWAYS', 
 'ОАО "ДОНАВИА"' : 'DONAVIA', 
 'ЗАО "АЗЕРБАЙДЖАН ХАВА ЙОЛЛАРЫ"' : 'AZERBAIJAN AIRLINES', 
 'LOT POLISH AIRLINES' : 'LOT POLISH AIRLINES', 
 'ALITALIA-LINEE AEREE ITALY' : 'ALITALIA', 
 'ЗАО "АВИАКОМПАНИЯ АЭРОСВИТ"' : 'AEROSVIT - UKRAINIAN AIRLINES', 
 'TURKISH AIRLINES' : 'TURKISH AIRLINES', 
 'HAINAN AIRLINES' : 'HAINAN AIRLINES', 
 'TAJIK AIR' : 'TAJIK AIR', 
 'ООО "АВИАПРЕДПРИЯТИЕ "СЕВЕРСТАЛЬ"' : 'SEVERSTAL', 
 'CYPRUS AIRWAYS LTD.' : 'CYPRUS AIRWAYS', 
 'НАЦИОНАЛЬНАЯ АК "ТУРКМЕНХОВАЕЛЛАРЫ"' : 'TURKMENISTAN AIRLINES', 
 'ОАО АВИАКОМПАНИЯ "ЯКУТИЯ"' : 'YAKUTIA AIRLINES', 
 'ОАО "ИЖАВИА"' : 'IZHAVIA', 
 'BULGARIA AIR' : 'BULGARIA AIR', 
 'ОАО "ПСКОВАВИА"' : 'PSKOVAVIA', 
 'NORWEGIAN AIR SHUTTLE' : 'NORWEGIAN', 
 'ООО "АВИАПРЕДПРИЯТИЕ "ГАЗПРОМАВИА"' : 'GAZPROMAVIA', 
 'AIR MOLDOVA' : 'AIR MOLDOVA', 
 'NOUVELAIR TUNISIE' : 'NOUVELAIR', 
 'AIR ONE' : 'AIR ONE', 
 'АНО "АЛЬЯНС АВИА"' : 'ALLIANCE AVIA', 
 'NEW LIVINGSTON SPA' : 'LIVINGSTON' , 
 'AVIA TRAFFIK COMPANY LLC' : 'AVIA TRAFFIC COMPANY', 
 'ОАО "РУСДЖЕТ"' : 'RUSJET', 
 'ЗАО АВИАЦИОННАЯ КОМПАНИЯ "РУСЛАЙН"' : 'RUSLINE', 
 'ЗАО "АВИА МЕНЕДЖМЕНТ ГРУП"' : 'AMG', 
 'ОАО "АВИАКОМПАНИЯ "ТАТАРСТАН"' : 'TATARSTAN AIRLINES', 
 'ЗАО "МИРНИНСКОЕ АВИАПРЕДПРИЯТИЕ АК "АЛРОСА"' : 'ALROSA', 
 'БАЛТИЙСКИЕ АВИАЛИНИИ' : 'BALTIC AIRLINES', 
 'ДЖЕТ ЭЙР ГРУПП' : 'JET AIR GROUP', 
 'ДЖЕТ ЭКСПРЕСС' : 'JET EXPRESS AIRLINES', 
 'КРЫЛЬЯ САМОТЛОРА' : 'WINGS OF SAMOTLOR', 
 'ЦСКБ-ПРОГРЕСС' : 'TSSKB-PROGRESS', 
 'АК "СИРИУС-АЭРО"' : 'SIRIUS AERO', 
 'ГАЗАВИА' : 'GAZAVIA', 
 'AVIES' : 'AVIES', 
 'EAST AIR' : 'EAST AIR', 
 'ООО "АРМАВИА"' : 'ARMAVIA', 
 'СПЕЦИАЛЬНЫЙ ЛЕТНЫЙ ОТРЯД "РОССИЯ"' : 'SPECIAL FLIGHT DETACHMENT "ROSSIYA"', 
 'JET FLITE' : 'JETFLITE', 
 'АБАКАН-АВИА' : 'ABAKAN-AVIA', 
 'АК "МЕРИДИАН"' : 'MERIDIAN', 
 'ООО "АВИАЦИОННАЯ КОМПАНИЯ "ЯМАЛ"' : 'YAMAL AIRLINES', 
 'НИЖНЕВАРТОВСКАВИА' : 'NIZHNEVARTOVSKAVIA', 
 'ЛЕТНЫЕ ПРОВЕРКИ И СИСТЕМЫ' : 'FLIGHT CHECKS AND SYSTEMS', 
 'CHARTER JETS' : 'CHARTER JETS', 
 'КОСТРОМСКОЕ АВИАПРЕДПРИЯТИЕ' : 'KOSTROMA AIR ENTERPRISE', 
 'TITAN AIRWAYS' : 'TITAN AIRWAYS', 
 'AIR ALSIE' : 'AIR ALSIE', 
 'EXECUJET EUROPE A/S' : 'EXECUJET', 
 'ФГУАП "МЧС РОССИИ"' : 'RUSSIAN MINISTER OF EMERGENCIES', 
 'ОАО "САРАТОВСКИЕ АВИАЛИНИИ"' : 'SARATOV AIRLINES', 
 '224 ЛЕТНЫЙ ОТРЯД' : '224TH FLIGHT UNIT', 
 'AIR FIX AVIATION' : 'AIRFIX AVIATION', 
 'AERO-CHARTER' : 'AERO-CHARTER', 
 'PROAIR AVIATION' : 'PROAIR AVIATION', 
 'MJET' : 'MJET', 
 'AIR UKRAINE INTERNATIONAL' : 'AIR UKRAINE INTERNATIONAL', 
 'AIR UKRAINE' : 'AIR UKRAINE', 
 'GAMA AVIATION' : 'GAMA AVIATION', 
 'УЧЕБНЫЕ ПОЛЕТЫ' : 'TRAINING FLIGHTS', 
 'АК "ЦЕНТР ЮГ"' : 'CENTER-SOUTH', 
 'ЮТЭЙР УКРАИНА' : 'UTAIR-UKRAINE', 
 'VISTA JET' : 'VISTA JET', 
 'AVCON JET AG' : 'AVCON JET', 
 'КОМИАВИАТРАНС' : 'KOMIAVIATRANS', 
 'SCANWINGS' : 'SCANWINGS', 
 'NET JETS TRANSPORTES AEREOS' : 'NETJETS', 
 'ATLAS JET HAVACILIC' : 'ATLASJET', 
 'ЗАО "ПЕТРОПАВЛОВСК"' : 'PETROPAVLOVSK-KAMCHATSKY AIR ENTERPRISE', 
 'ООО "АВИАКОМПАНИЯ "ТУЛПАР"' : 'TULPAR AIR', 
 'HAMBURG AIRWAYS' : 'HAMBURG AIRWAYS', 
 'ОАО АК "АВИАКОН ЦИТОТРАНС"' : 'AVIACON ZITOTRANS', 
 'АК "БЫЛИНА"' : 'BYLINA', 
 'ОАО "АВИАКОМПАНИЯ "АТРАН"' : 'ATRAN', 
 'EUROPE AIRPOST' : 'EUROPE AIRPOST', 
 'AVIONES UNIDOS, S.A. DE C.V.' : 'AVIONES UNIDOS', 
 'LUX AVIATION' : 'LUX AVIATION', 
 'MHS AVIATION GMBH' : 'MHS AVIATION', 
 'DC-AVIATION' : 'DC-AVIATION', 
 'VUELING AIRLINES' : 'VUELING', 
 'IBERIA LINEAS AEREAS DE E' : 'IBERIA', 
 'PANAVIATIC' : 'PANAVIATIC', 
 'ЗАО ПО "КОСМОС" РКК ЭНЕРГИЯ ИМ.С.П.КОРОЛЕВА' : 'KOSMOS AIRLINES', 
 'SERVIZI AEREI' : 'SERVIZI AEREI', 
 'FINNISH AIR FORCE' : 'FINNISH AIR FORCE', 
 'ETIHAD AIRWAYS' : 'ETIHAD AIRWAYS', 
 'ROYAL JORDANIAN' : 'ROYAL JORDANIAN', 
 'SINGAPORE AIRLINES' : 'SINGAPORE AIRLINES', 
 'QANTAS' : 'QANTAS', 
 'TAP AIR PORTUGAL' : 'TAP AIR PORTUGAL', 
 'ООО "ЮТЭЙР-ЭКСПРЕCC"' : 'UTAIR-EXPRESS', 
 '2 АРХАНГЕЛЬСКИЙ ОАО' : '2ND ARKHANGELSK UNITED AVIATION DIVISION', 
 'AIR MALTA' : 'AIR MALTA', 
 'JAT AIRWAYS' : 'JAT AIRWAYS', 
 'ASTRA AIRLINES' : 'ASTRA AIRLINES', 
 'TUNIS AIR' : 'TUNIS AIR', 
 'AEGEAN' : 'AEGEAN', 
 'MONTENEGRO AIRLINES' : 'MONTENEGRO AIRLINES', 
 'AIR ONIX' : 'AIR ONIX', 
 'TRAVEL SERVICE' : 'TRAVEL SERVICE', 
 'ELITAVIA D.O.O.' : "ELIT'AVIA", 
 'UKRAINE AIR ALLIANCE' : 'UKRAINE AIR ALLIANCE', 
 'EUROP-STAR AIRCRAFT GMBH' : 'EUROP STAR AIRCRAFT', 
 'TRADEAIR' : 'TRADEAIR', 
 'JET2.COM LIMITED' : 'JET2.COM', 
 'МЕРИДИАН ПЛЮС' : 'MERIDIAN-PLUS', 
 'ГРОЗНЫЙ АВИА' : 'GROZNY AVIA', 
 'АВИАКОМПАНИЯ "ЕВРО-АЗИЯ ЭЙР"' : 'EURO-ASIA AIR', 
 'ГУП "АЭРОПОРТ ОРЕНБУРГ"' : 'ORENBURG AIRPORT', 
 'BLUE BIRD AIRWAYS' : 'BLUEBIRD AIRWAYS', 
 'ООО АК "АВИАСТАР-ТУ"' : 'AVIASTAR-TU', 
 'ЗАО "ИРАЭРО"' : 'IRAERO', 
 'FLYBE' : 'FLYBE', 
 'CZECH AIR FORCE' : 'CZECH AIR FORCE', 
 'АВИАКОМПАНИЯ "МОСКОВИЯ"' : 'MOSKOVIA AIRLINES', 
 'BLUE PANORAMA AIRLINES' : 'BLUE PANORAMA AIRLINES', 
 'ООО АК "БАРКОЛ"' : 'BARKOL', 
 'ICELANDAIR' : 'ICELANDAIR', 
 'СЕВЕРО-ЗАПАДНАЯ БАЗА АВИАЛЕСООХРАНЫ' : 'NORTH-WESTERN AVIATION FOREST PROTECTION', 
 'SOMON AIR' : 'SOMON AIR', 
 'AIR HAMBURG PRIVATE JETS' : 'AIR HAMBURG PRIVATE JETS', 
 'БИЗНЕС АЭРО' : 'BUSINESS AERO', 
 'АВИАЛИНИИ МОРДОВИИ' : 'MORDOVIA AIRLINES', 
 'GERMAN AIR FORCE' : 'GERMAN AIR FORCE', 
 'АЭРОЛИМУЗИН' : 'AEROLIMOUSINE', 
 'IBERIA EXPRESS' : 'IBERIA EXPRESS', 
 'ВВС РОССИИ' : 'RUSSIAN AIR FORCE', 
 'MISTRAL AIR' : 'MISTRAL AIR', 
 'GEORGIAN AIRWAYS' : 'GEORGIAN AIRWAYS', 
 'АМУРСКАЯ БАЗА АВИАЛЕСООХРАНЫ' : 'AMUR AVIATION FOREST PROTECTION' , 
 '223 ЛЁТНЫЙ ОТРЯД' : '223RD FLIGHT UNIT', 
 'ART AVIATION' : 'ART AVIATION', 
 'HYPERION AVIATION LTD' : 'HYPERION AVIATION', 
 'MAHAN AIRLINES' : 'MAHAN AIRLINES', 
 'ЗАО "КОНВЕРС АВИА"' : 'CONVERS AVIA', 
 'SHAR INC LTD' : 'SHAR INK', 
 'FAI RENT-A-JET' : 'FAI RENT-A-JET', 
 'ООО  "ВОЛГА-ДНЕПР"' : 'VOLGA-DNEPR', 
 'PRIVILEGE STYLE' : 'PRIVILEGE STYLE', 
 'EASTERN SKY JETS' : 'EASTERN SKY JETS', 
 'UNITED STATES AIR FORCE' : 'UNITED STATES AIR FORCE', 
 'FORCA AEREA BRASILEIRA' : 'FORCA AEREA BRASILEIRA', 
 'ETHIOPIAN AIRLINES' : 'ETHIOPIAN AIRLINES', 
 'JAPAN AIR FORCE' : 'JAPAN AIR FORCE', 
 'ALL NIPPON AIRWAYS' : 'ALL NIPPON AIRWAYS', 
 'AMI AERONAUTICA MILITARE' : 'AMI AERONAUTICA MILITARE', 
 'AIR CHINA LTD.' : 'AIR CHINA', 
 'CANADIAN AIR FORCE' : 'CANADIAN AIR FORCE', 
 'AUGUSTA AIR LUFTFAHRTUNTERNEHMEN' : 'AUGUSTA AIR', 
 "ARMEE DE L'AIR FRANCAISE" : "ARMEE DE L'AIR FRANCAISE", 
 'ООО АК "ВИМ-АВИА"' : 'VIM AIRLINES', 
 'ADRIA AIRWAYS' : 'ADRIA AIRWAYS', 
 'AIR LITUANICA' : 'AIR LITUANICA', 
 'EGYPTAIR' : 'EGYPTAIR', 
 'XCLUSIVE JET CHARTER' : 'XCLUSIVE JET CHARTER', 
 'AIR MEDITERRANEE' : 'AIR MEDITERRANEE', 
 'ЗАО "КАПО АВИА"' : 'KAPO AVIA', 
 'SATA INTERNATIONAL' : 'SATA INTERNATIONAL', 
 'ООО "АВИАКОМПАНИЯ "ИКАР"' : 'PEGAS FLY', 
 'TRANSAVIA' : 'TRANSAVIA', 
 'WIND ROSE AIRLINES' : 'WINDROSE AIRLINES', 
 'AIR ARMENIA' : 'AIR ARMENIA', 
 'FARNAIR' : 'FARNAIR', 
 'SOSOLISO AIRLINES' : 'SOSOLISO AIRLINES', 
 'PROGRESS MULTITRADE CO. - PMT AIR' : 'PMT AIR', 
 'ОАО НПП "МИР" (АВИАПРОМ РОССИИ)' : 'MIR - RUSSIAN AVIAPROM', 
 'ООО "АК АЙ ФЛАЙ"' : 'I-FLY', 
 'CATHAY PACIFIC AIRWAYS LTD' : 'CATHAY PACIFIC', 
 'ОАО "АК "ВОСТОК"' : 'VOSTOK AIRLINES'}

In [42]:
# replace the airline names using the dictionary
df.replace({'airline_name': airlines}, inplace=True)

In [43]:
# check the value percentages in the 'aircraft_type_code' column
df.aircraft_type_code.value_counts(normalize=True)

32S    0.249416
320    0.116012
735    0.075104
A81    0.070699
319    0.059138
         ...   
I14    0.000015
721    0.000015
L29    0.000007
K32    0.000007
747    0.000007
Name: aircraft_type_code, Length: 174, dtype: float64

In [44]:
# check the value percentages in the 'aircraft_type_name' column
df.aircraft_type_name.value_counts(normalize=True)

AIRBAS INDUSTRY            0.249416
AIRBUS INDUSTRIE           0.235510
BOEING                     0.221962
ANTONOV                    0.079661
BOMBARDIER REGIONAL JET    0.058286
BOMBARDIER                 0.018951
EMBRAER                    0.017655
MIL                        0.012049
DASSAULT                   0.011030
CESSNA                     0.009464
AEROSPATIALE/ALENIA        0.008387
HAWKER BEECHCRAFT          0.008110
GULFSTREAM AEROSPACE       0.008110
LEARJET                    0.007979
FOKKER                     0.007950
TUPOLEV                    0.007550
SAAB                       0.006800
DE HAVILLAND               0.005504
YAKOVLEV                   0.005460
EUROCOPTER                 0.004900
AGUSTA                     0.003873
ROBINSON                   0.002774
ILYUSHIN                   0.001864
KAMOV                      0.001529
MCDONNELL DOUGLAS          0.001456
BEECHCRAFT                 0.000641
AGUSTA WESTLAND            0.000561
AEROSPATIALE               0

**Aircraft manufacturers** have some repetitions and incorrect names, they will be corrected.  
<br/><br/>

In [45]:
# replace incorrect aircraft manufacturers names
df['aircraft_type_name'].replace(['AIRBAS INDUSTRY', 'AIRBUS INDUSTRIE', 'BOMBARDIER REGIONAL JET'],
                                 ['AIRBUS', 'AIRBUS', 'BOMBARDIER'], inplace=True)

In [46]:
# display the most common aircraft manufacturers
df.aircraft_type_name.value_counts(normalize=True).head(15)

AIRBUS                  0.484926
BOEING                  0.221962
ANTONOV                 0.079661
BOMBARDIER              0.077237
EMBRAER                 0.017655
MIL                     0.012049
DASSAULT                0.011030
CESSNA                  0.009464
AEROSPATIALE/ALENIA     0.008387
HAWKER BEECHCRAFT       0.008110
GULFSTREAM AEROSPACE    0.008110
LEARJET                 0.007979
FOKKER                  0.007950
TUPOLEV                 0.007550
SAAB                    0.006800
Name: aircraft_type_name, dtype: float64

In [47]:
# check the value percentages in the 'service_type' column
df.service_type.value_counts(normalize=True)

Normal Service (Sheduled, Passenger)                                0.876089
Business Aviation/Air Taxi (Others, Not Specific)                   0.047119
Passenger Only (Charter, Passenger)                                 0.038465
Helicopters (All flights)                                           0.015090
General Aviation (Others, Not specific)                             0.005883
Правительство РФ (ВИП)                                              0.004552
Non-revenue (Positioning/Ferry/Delivery/Demo; Not specific)         0.003927
Cargo and/or Mail (Charter, Cargo/Mai)                              0.003016
Normal Service (Additional Flights, Passenger)                      0.003001
Loose loaded cargo and/or preloaded devices(Sheduled,Cargo/Mail)    0.000901
Technical Test (Others, Not specific)                               0.000650
Правительство РФ (Правительства других государств)                  0.000538
State order (Гос.задание)                                           0.000297

Some **service type** names are in the Cyrillic alphabet, they will be translated into English and cleaned up.  
<br/><br/>

In [48]:
# replace service type names
df['service_type'].replace(['Правительство РФ (ВИП)', 'Правительство РФ (Правительства других государств)',
                            'State order (Гос.задание)'], ['Government of the Russian Federation (VIP)',
                            'Government of the Russian Federation (Governments of other states)',
                            'State Order'], inplace=True)

In [49]:
# check the value percentages in the 'domestic' column
df.domestic.value_counts(normalize=True)

Domestic         0.593685
International    0.406315
Name: domestic, dtype: float64

In [50]:
# confirm the sum of the different values
df.domestic.value_counts().sum()

195292

In [51]:
# the 'domestic' column has 1 empty value
# check which row has the empty value, using a condition
d2 = df[df['domestic'].isnull()>0]
# check the 'country_name' column
d2['country_name']

17951    FRANCE
Name: country_name, dtype: object

In [52]:
# this is an international flight
# set the correct value for the empty cell
df.loc[df['domestic'].isnull()>0, 'domestic'] = 'International'

In [53]:
# confirm the sum of the different values
df.domestic.value_counts().sum()

195293

In [54]:
# convert 'domestic' column into binary
df.domestic[df.domestic == 'Domestic'] = 1
df.domestic[df.domestic == 'International'] = 0

In [55]:
# check the value percentages in the 'cancelled' column
df.cancelled.value_counts(normalize=True)

Y    1.0
Name: cancelled, dtype: float64

In [56]:
# check the sum of the different values
df.cancelled.value_counts().sum()

89

In [57]:
# convert 'cancelled' column into binary
df.cancelled.fillna(0, inplace=True)
df.cancelled[df.cancelled == 'Y'] = 1

In [58]:
# check the value percentages in the 'cancelled' column
df.cancelled.value_counts(normalize=True)

0    0.999544
1    0.000456
Name: cancelled, dtype: float64

In [59]:
# confirm the sum of the different values
df.cancelled.value_counts().sum()

195293

In [60]:
# check the value percentges in the 'regular' column
df.regular.value_counts(normalize=True)

Regular        0.91759
Non-Regular    0.08241
Name: regular, dtype: float64

In [61]:
# confirm the sum of the different values
df.regular.value_counts().sum()

195293

In [62]:
# convert the 'regular' column into binary
df.regular[df.regular == 'Regular'] = 1
df.regular[df.regular == 'Non-Regular'] = 0
df.head()

date  year  month  day  weekday  hour departing terminal_num  \
0  2013-01-01 03:00  2013      1    1        1     3         0            2   
1  2013-01-01 03:00  2013      1    1        1     3         0            1   
2  2013-01-01 05:00  2013      1    1        1     5         1            2   
3  2013-01-01 06:00  2013      1    1        1     6         1            1   
4  2013-01-01 07:00  2013      1    1        1     7         0            2   

  airline_code           airline_name  ... flight_count load_factor  \
0           N4      NORDWIND AIRLINES  ...            1      67.586   
1           OZ        ASIANA AIRLINES  ...            1         NaN   
2           N4      NORDWIND AIRLINES  ...            1      94.667   
3           OZ        ASIANA AIRLINES  ...            1         NaN   
4           LY  EL AL ISRAEL AIRLINES  ...            1       8.235   

  baggage_weight mail_weight temperature relative_humidity  \
0         3327.0         NaN         0.9                94   
1            NaN         NaN         0.9                94   
2         3224.0         0.0         0.9                94   
3            NaN         NaN         0.7                94   
4          244.0       173.0         0.8                94   

  apparent_temperature precipitation weather_code wind_speed  
0                 -2.4           0.0            3       13.0  
1                 -2.4           0.0            3       13.0  
2                 -2.9           0.0            3       17.0  
3                 -3.3           0.0            3       18.1  
4                 -3.6           0.2           71       19.6  

[5 rows x 40 columns]

In [63]:
# check the value percentages in the 'CIS_country' column
df.CIS_country.value_counts(normalize=True)

Y    1.0
Name: CIS_country, dtype: float64

In [64]:
# check the values in the 'CIS_country' column
df.CIS_country.value_counts()

Y    20733
Name: CIS_country, dtype: int64

The **'CIS_country'** column should represent a binary value and has a lot of empty cells.  
It should probably refer to flights to/from one of the 8 official CIS countries: Armenia, Azerbaijan, Belarus, Kazakhstan, Kyrgyzstan, Moldova, Tajikistan, Uzbekistan.  
<br/><br/>

In [65]:
# verify the above assumption
df[df.CIS_country == 'Y']['country_name'].value_counts()

UKRAINE                 7371
UZBEKISTAN              4078
KAZAKHSTAN              2263
BELARUS                 1791
TAJIKISTAN              1588
AZERBAIJAN              1084
KYRGYZSTAN              1022
ARMENIA                  713
MOLDOVA, REPUBLIC OF     603
TURKMENISTAN             216
GEORGIA                    4
Name: country_name, dtype: int64

In [66]:
# convert the 'CIS_country' column into binary
df.CIS_country.fillna(0, inplace=True)
df.CIS_country[df.CIS_country == 'Y'] = 1

In [67]:
# check the value percentages in the 'CIS_country' column
df.CIS_country.value_counts(normalize=True)

0    0.893836
1    0.106164
Name: CIS_country, dtype: float64

In [68]:
# confirm the sum of different values
df.CIS_country.value_counts().sum()

195293

In [69]:
# check the value percentages in the 'country_name' column
df.country_name.value_counts(normalize=True)

RUSSIAN FEDERATION                   0.523065
RUSSIA EAST OF THE URAL MOUNTAINS    0.063484
GERMANY                              0.041333
UKRAINE                              0.037743
ITALY                                0.022223
                                       ...   
BRAZIL                               0.000005
KENYA                                0.000005
CAMBODIA                             0.000005
GRENADA                              0.000005
SEYCHELLES                           0.000005
Name: country_name, Length: 88, dtype: float64

In [70]:
# check the value percentages in the 'flight_number' column
df.flight_number.value_counts(normalize=True)

FV  341    0.001864
KL 1396    0.001859
FV  574    0.001859
KL 1395    0.001859
FV  641    0.001859
             ...   
OK 1884    0.000005
OK 1885    0.000005
VPC6064    0.000005
FJE188C    0.000005
SU 1796    0.000005
Name: flight_number, Length: 5660, dtype: float64

In [71]:
# check the sum of different values
df.flight_number.value_counts().sum()

195289

In [67]:
# display the number of different scheduled flights
len(list(df['flight_number'].unique()))

5661

The **'main_delay_code'** and **'source_delay'** columns have been checked already.  
A binary column will be created for modelling purposes.  
<br/><br/>

In [69]:
# check the null values in the 'main_delay_code' column
df['main_delay_code'].isnull().sum()

162033

In [72]:
# check the value percentages in the 'main_delay_code' column
df['main_delay_code'].value_counts(normalize=True)

93     0.317108
99     0.137462
89     0.075135
3      0.057005
85     0.049880
         ...   
27     0.000060
25     0.000060
60     0.000060
93W    0.000030
56     0.000030
Name: main_delay_code, Length: 88, dtype: float64

In [73]:
# display the unique values
df['main_delay_code'].unique()

array([nan, '89', '86', '15', '75', '93', '4', '46', '3', '14', '85',
       '99', '36', '10', '72', '18', '63', '41', '11', '32', '5', '47',
       '96', '81', '87', '34', '77', '65', '31', '1', '95', '91', '43',
       '33', '38', '39', '16', '29', '97', '66', '42', '19', '37', '6',
       '94', '68', '88', '12', '2', '92', '9', '61', '76', '82', '51',
       '23', '69', '55', '71', '35', '8', '57', '62', '22', '67', '21',
       '70', '25', '0', '98', '13', '27', '45', '58', '52', '28', '64',
       '83', '49', '17', '84', '44', '73', '60', '93W', '48', '24', '56',
       '93Y'], dtype=object)

In [75]:
# check the number of unique values
len(list(df['main_delay_code'].unique()))

89

In [76]:
# replace nan value with None
df['main_delay_code'] = df['main_delay_code'].replace(np.nan, None)

In [77]:
# display the unique values
df['main_delay_code'].unique()

array([None, '89', '86', '15', '75', '93', '4', '46', '3', '14', '85',
       '99', '36', '10', '72', '18', '63', '41', '11', '32', '5', '47',
       '96', '81', '87', '34', '77', '65', '31', '1', '95', '91', '43',
       '33', '38', '39', '16', '29', '97', '66', '42', '19', '37', '6',
       '94', '68', '88', '12', '2', '92', '9', '61', '76', '82', '51',
       '23', '69', '55', '71', '35', '8', '57', '62', '22', '67', '21',
       '70', '25', '0', '98', '13', '27', '45', '58', '52', '28', '64',
       '83', '49', '17', '84', '44', '73', '60', '93W', '48', '24', '56',
       '93Y'], dtype=object)

In [78]:
# check the number of unique values
len(list(df['main_delay_code'].unique()))

89

In [79]:
# create a new binary column based on condition
df['delayed'] = [0 if x == None else 1 for x in df['main_delay_code']]

In [80]:
# check the value percentages in the new binary colunm
df.delayed.value_counts()

0    162033
1     33260
Name: delayed, dtype: int64

In [81]:
# confirm the sum of the different values 
df.delayed.value_counts().sum()

195293

In [83]:
# check the value percentages in the 'runway' column
df.runway.value_counts(normalize=True)

28R    0.333625
28L    0.266331
10L    0.217097
10R    0.160870
H      0.022076
Name: runway, dtype: float64

There are two runways, used in both directions. H represents the heliport.  
<br/><br/>

In [84]:
# check the value percentage for the 'spot_number' column, that represent the aircraft parking slot
df.spot_number.value_counts(normalize=True)

66     0.032512
65     0.031420
67     0.030845
35     0.029906
36     0.029163
         ...   
249    0.000007
275    0.000007
220    0.000007
221    0.000007
211    0.000007
Name: spot_number, Length: 286, dtype: float64

### 2.1 Passengers columns

In [89]:
# display summary statistics for the passengers columns
df[['pax_total', 'pax_1', 'pax_2', 'pax_3', 'pax_4', 'pax_5', 'pax_6']].describe(include='all')

pax_total          pax_1          pax_2          pax_3  \
count  123018.000000  195293.000000  195293.000000  195293.000000   
mean      103.844649      62.046279       3.360161       0.720062   
std        65.204536      66.526937       8.729913       2.025564   
min         0.000000       0.000000       0.000000       0.000000   
25%        63.000000       0.000000       0.000000       0.000000   
50%        98.000000      54.000000       0.000000       0.000000   
75%       130.000000     107.000000       3.000000       1.000000   
max      1002.000000     999.000000     425.000000     337.000000   

               pax_4          pax_5          pax_6  
count  195293.000000  195293.000000  195293.000000  
mean        0.046453       2.024307      56.529748  
std         1.152120       5.212153      69.956528  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000  
50%         0.000000       0.000000      38.000000  
75%         0.000000       2.000000     103.000000  
max       298.000000     479.000000     605.000000

In [90]:
# check the value percentages for the 'pax_total' column
df.pax_total.value_counts(normalize=True)

75.0     0.011356
0.0      0.010966
128.0    0.009600
126.0    0.009560
124.0    0.009486
           ...   
389.0    0.000008
423.0    0.000008
453.0    0.000008
408.0    0.000008
370.0    0.000008
Name: pax_total, Length: 497, dtype: float64

In [91]:
# check the value percentages for the 'pax_1' column
df.pax_1.value_counts(normalize=True)

0      0.377192
116    0.005694
121    0.005663
106    0.005633
110    0.005597
         ...   
353    0.000005
499    0.000005
330    0.000005
490    0.000005
600    0.000005
Name: pax_1, Length: 498, dtype: float64

In [92]:
# check the value percentages for the 'pax_2' column
df.pax_2.value_counts(normalize=True)

0      0.556461
1      0.095585
2      0.071421
3      0.052178
4      0.037364
         ...   
137    0.000005
134    0.000005
149    0.000005
131    0.000005
200    0.000005
Name: pax_2, Length: 143, dtype: float64

In [93]:
# check the value percentages for the 'pax_3' column
df.pax_3.value_counts(normalize=True)

0      0.697864
1      0.142698
2      0.070561
3      0.035070
4      0.019478
5      0.011721
6      0.007415
7      0.004808
8      0.003016
9      0.002099
10     0.001475
11     0.000942
12     0.000830
13     0.000589
14     0.000312
15     0.000271
16     0.000159
17     0.000143
19     0.000123
18     0.000118
21     0.000087
20     0.000077
24     0.000020
26     0.000020
22     0.000015
31     0.000010
92     0.000010
23     0.000010
25     0.000010
337    0.000005
45     0.000005
123    0.000005
30     0.000005
78     0.000005
291    0.000005
32     0.000005
59     0.000005
230    0.000005
Name: pax_3, dtype: float64

In [94]:
# check the value percentages for the 'pax_4' column
df.pax_4.value_counts(normalize=True)

0      0.990420
1      0.002181
2      0.001833
8      0.001342
3      0.001004
4      0.000881
7      0.000553
5      0.000476
6      0.000451
12     0.000389
9      0.000133
11     0.000128
10     0.000123
14     0.000020
13     0.000015
298    0.000005
259    0.000005
18     0.000005
16     0.000005
118    0.000005
15     0.000005
48     0.000005
147    0.000005
32     0.000005
102    0.000005
Name: pax_4, dtype: float64

In [95]:
# check the value percentages for the 'pax_5' column
df.pax_5.value_counts(normalize=True)

0      0.684085
1      0.051364
2      0.046325
3      0.035844
4      0.029873
         ...   
49     0.000005
59     0.000005
112    0.000005
70     0.000005
56     0.000005
Name: pax_5, Length: 104, dtype: float64

In [96]:
# check the value percentages for the 'pax_6' column
df.pax_6.value_counts(normalize=True)

0      0.460923
120    0.014675
119    0.007691
75     0.006447
118    0.006411
         ...   
332    0.000005
339    0.000005
432    0.000005
445    0.000005
324    0.000005
Name: pax_6, Length: 473, dtype: float64

There are seven different **passenger** columns.
I assume from their names that one represents the total of passengers, while the others are used for passengers classes or different passengers typologies (e.g. crew members, navigators...).  
However, there are not enough information to understand what they really represent, and the summary statistics shows some figures that seem unrealistic.  
<br/><br/>

In [97]:
# check the value pecentages for the 'flight_count' column
df.flight_count.value_counts(normalize=True)

1    1.0
Name: flight_count, dtype: float64

In [98]:
# confirm the number of values for the 'flight_count' column
df.flight_count.value_counts()

1    195293
Name: flight_count, dtype: int64

In [99]:
# display the current size of the dataframe
df.shape

(195293, 41)

In [100]:
# the 'flight_count' column is a binary variable, but its value is always equal to 1 so it could be dropped
# drop the column
df.drop(columns=['flight_count'], inplace=True)
# display the new size of the dataframe
df.shape

(195293, 40)

### 2.2 Load Factor column

In [101]:
# check the value percentages for the 'load_factor' column
df.load_factor.value_counts(normalize=True)

100.000    0.342461
0.000      0.005207
75.000     0.005180
66.667     0.004703
50.000     0.004637
             ...   
40.708     0.000005
273.636    0.000005
192.982    0.000005
193.860    0.000005
52.857     0.000005
Name: load_factor, Length: 4162, dtype: float64

In [102]:
# check the sum of the values
df.load_factor.value_counts().sum()

182447

The **load factor** is a percentage, so it should never be greater than 100%.  
I assume the scale goes from 0 to 100, so there are apparently incorrect values.  
<br/><br/>

In [103]:
# check the number of flights with load factor greater than 100%
df[df.load_factor > 100]

date  year  month  day  weekday  hour departing  \
6       2013-01-01 09:00  2013      1    1        1     9         1   
11      2013-01-01 10:00  2013      1    1        1    10         1   
36      2013-01-01 14:00  2013      1    1        1    14         1   
75      2013-01-01 17:00  2013      1    1        1    17         1   
81      2013-01-01 18:00  2013      1    1        1    18         1   
...                  ...   ...    ...  ...      ...   ...       ...   
195159  2013-12-31 16:00  2013     12   31        1    16         1   
195164  2013-12-31 16:00  2013     12   31        1    16         1   
195249  2013-12-31 19:00  2013     12   31        1    19         1   
195258  2013-12-31 19:00  2013     12   31        1    19         0   
195290  2013-12-31 22:00  2013     12   31        1    22         0   

       terminal_num airline_code               airline_name  ... load_factor  \
6                 1           R2                    ORENAIR  ...     110.588   
11                1           FV           ROSSIYA AIRLINES  ...     110.897   
36                2           7K                   METROJET  ...     114.211   
75                2           OS          AUSTRIAN AIRLINES  ...     103.846   
81                2           SK  SAS-SCANDINAVIAN AIRLINES  ...     100.833   
...             ...          ...                        ...  ...         ...   
195159            2           LX                      SWISS  ...     103.205   
195164            2           PS  AIR UKRAINE INTERNATIONAL  ...     122.807   
195249            2           KL   KLM-ROYAL DUTCH AIRLINES  ...     100.588   
195258            1           S7                S7 AIRLINES  ...     108.947   
195290            1           U6              URAL AIRLINES  ...     104.211   

       baggage_weight mail_weight temperature relative_humidity  \
6              1847.0         NaN         1.0                92   
11             1592.0         NaN         0.9                93   
36             2062.0         0.0         1.4                96   
75             1905.0         NaN         1.8                97   
81             1439.0         NaN         2.1                95   
...               ...         ...         ...               ...   
195159         1878.0         0.0         2.6                86   
195164         1117.0         NaN         2.6                86   
195249            NaN         NaN         1.7                85   
195258         1456.0         0.0         1.7                85   
195290         2044.0         0.0         1.2                84   

       apparent_temperature precipitation weather_code wind_speed delayed  
6                      -3.7           0.1           71       21.0       0  
11                     -3.7           0.4           73       20.5       0  
36                     -2.8           0.5           71       18.4       0  
75                     -2.0           0.4           51       17.7       1  
81                     -1.6           0.2           51       17.1       1  
...                     ...           ...          ...        ...     ...  
195159                 -0.8           0.0            3       15.1       0  
195164                 -0.8           0.0            3       15.1       0  
195249                 -2.2           0.0            2       16.3       0  
195258                 -2.2           0.0            2       16.3       0  
195290                 -2.8           0.0            2       16.1       1  

[7737 rows x 40 columns]

In [104]:
# display the mean value for the load factor
df.load_factor.mean()

81.16345147357833

There are **almost 8,000** flights with a load factor greater than 100%, which represents **3.9%** of the total number of flights.  
This could be a data entry error, but it could also be downright bad business practice (assuming that load factor accounts for passengers, baggage and mail).  
At this stage, I don't have enough information to decide whether to impute a different load factor to these flights or to remove them altogether.  
<br/><br/>

In [105]:
# check the value percentages for the 'baggage_weight' column
df.baggage_weight.value_counts(normalize=True)

0.0       0.015815
650.0     0.000917
493.0     0.000917
468.0     0.000859
600.0     0.000859
            ...   
3389.0    0.000008
3308.0    0.000008
4939.0    0.000008
4053.0    0.000008
9405.0    0.000008
Name: baggage_weight, Length: 5730, dtype: float64

In [106]:
# display the mean value for the baggage weight
df.baggage_weight.mean()

1134.466371009035

In [107]:
# check the value percentages for the 'mail_weight' column
df.mail_weight.value_counts(normalize=True)

0.0       0.685672
1.0       0.007100
4.0       0.003532
2.0       0.003508
19.0      0.003460
            ...   
1875.0    0.000012
769.0     0.000012
631.0     0.000012
951.0     0.000012
1471.0    0.000012
Name: mail_weight, Length: 1164, dtype: float64

In [108]:
# display the mean value for the mail weight
df.mail_weight.mean()

50.70861158658598

In [109]:
# display summary statistics for the weather columns
df[['temperature', 'relative_humidity', 'apparent_temperature', 'precipitation', 'weather_code', 'wind_speed']].describe()

temperature  relative_humidity  apparent_temperature  precipitation  \
count  195293.000000      195293.000000         195293.000000  195293.000000   
mean        9.103383          78.248683              6.827355       0.084761   
std         9.662356          15.507387             11.554376       0.317358   
min       -25.200000          29.000000            -32.900000       0.000000   
25%         2.200000          67.000000             -1.600000       0.000000   
50%        10.200000          82.000000              7.500000       0.000000   
75%        17.000000          91.000000             16.100000       0.000000   
max        30.600000         100.000000             34.500000       5.600000   

        weather_code     wind_speed  
count  195293.000000  195293.000000  
mean       11.385155      11.967680  
std        21.362831       5.552934  
min         0.000000       0.000000  
25%         1.000000       7.700000  
50%         2.000000      11.400000  
75%         3.000000      15.500000  
max        73.000000      37.800000

In [110]:
# display the value percentages for the 'weather_code' column
df.weather_code.value_counts(normalize=True)

3     0.289928
1     0.202071
0     0.188389
2     0.143379
51    0.086557
53    0.034082
71    0.030733
61    0.013073
73    0.008367
63    0.003421
Name: weather_code, dtype: float64

In [111]:
# check for missing values
df.isnull().sum()

date                         0
year                         0
month                        0
day                          0
weekday                      0
hour                         0
departing                    0
terminal_num             59864
airline_code             11161
airline_name             11353
aircraft_type_code       57936
aircraft_type_name       57936
service_type                 0
domestic                     0
cancelled                    0
regular                      0
CIS_country                  0
country_name                 0
flight_number                4
main_delay_code         162033
source_delay            180400
runway                  126803
spot_number              57927
pax_total                72275
pax_1                        0
pax_2                        0
pax_3                        0
pax_4                        0
pax_5                        0
pax_6                        0
load_factor              12846
baggage_weight           74207
mail_wei

In [112]:
# encode binary columns as number
df['departing'] = df['departing'].astype(float)
df['domestic'] = df['domestic'].astype(float)
df['cancelled'] = df['cancelled'].astype(float)
df['regular'] = df['regular'].astype(float)
df['CIS_country'] = df['CIS_country'].astype(float)

In [113]:
# check data types for the different columns
df.dtypes

date                     object
year                      int64
month                     int64
day                       int64
weekday                   int64
hour                      int64
departing               float64
terminal_num             object
airline_code             object
airline_name             object
aircraft_type_code       object
aircraft_type_name       object
service_type             object
domestic                float64
cancelled               float64
regular                 float64
CIS_country             float64
country_name             object
flight_number            object
main_delay_code          object
source_delay             object
runway                   object
spot_number              object
pax_total               float64
pax_1                     int64
pax_2                     int64
pax_3                     int64
pax_4                     int64
pax_5                     int64
pax_6                     int64
load_factor             float64
baggage_

In [114]:
# save the modified dataframe as CSV
df.to_csv('datasets/mod/df_01.csv', encoding='utf-8', index=False)